lib/HooksHandler.php

```diff
@@ -257,7 +257,10 @@ public function savePasswordForCreatedUser(GenericEvent $event) {
 	/**
 	 * Flags the session to require a password change
 	 */
-	protected function forcePasswordChange() {
+	protected function forcePasswordChange($firstLogin = null, IUser $user = null) {
+		if (($firstLogin === true) && ($user !== null)) {
+			$this->config->setUserValue($user->getUID(), 'password_policy', 'firstLoginPasswordChange', '1');
```

Comment:
you likely don't need this and check the existing value of "last_login" in the user object / accounts table

Reply:
if last_login is null it means the user has never logged in before

Reply:
Initially I thought to use the same. But then I believe when user tries to login with username + password from the webUI,  the code reaches
- https://github.com/owncloud/core/blob/master/lib/private/User/Session.php#L520
- https://github.com/owncloud/core/blob/master/lib/private/User/Manager.php#L254
And hence the "last_login" will be updated. So if we try to use "last_login" the value will be a positive integer. ( Cross verified just now ).

Reply:
Do you mean that even with failed login we do set the "last_login" value ? That doesn't sound right as it would likely also trigger skeleton creation.

Mind debugging to confirm, in case there's another hidden bug ?

Reply:
Sure I can debug this. But my comment above ( https://github.com/owncloud/password_policy/pull/81#discussion_r207847148) wasn't with respect to failed login. The scenario I had was like this:
- enforced first login password change in password policy settings
- Create a user `user1`
- Try to login from web ui as `user1`.
- This would populate "last_login" by the time code reaches https://github.com/owncloud/password_policy/blob/master/lib/HooksHandler.php#L317

Reply:
The skeleton is copied by the time password_policy code reaches https://github.com/owncloud/password_policy/blob/master/lib/HooksHandler.php#L317

Reply:
I think the behaviour with the "last_login" is ok with respect to my comments. Because the "last_login" will be updated https://github.com/owncloud/core/blob/master/lib/private/User/Session.php#L529. Since its the first time login and its validated, the code reaches https://github.com/owncloud/core/blob/master/lib/private/User/Session.php#L532 followed by https://github.com/owncloud/core/blob/master/lib/private/User/Session.php#L465. By this time the skeleton is already copied. 

Now password_policy triggers the password reset template. So "last_login" value is already updated for the user.

So I believe the code behaviour is correct.

Reply:
Okay, thanks for the details.

In that case the proposed approach makes sense.